mlrun/api/initial_data.py

```diff
@@ -508,33 +539,147 @@ def _rename_marketplace_kind_to_hub(
         db._upsert(db_session, [hub], ignore=True)
 
 
-def _perform_version_1_data_migrations(
+def _perform_version_4_data_migrations(
     db: mlrun.api.db.sqldb.db.SQLDB, db_session: sqlalchemy.orm.Session
 ):
-    _enrich_project_state(db, db_session)
-    _fix_artifact_tags_duplications(db, db_session)
-    _fix_datasets_large_previews(db, db_session)
+    _migrate_artifacts_table_v2(db, db_session)
 
 
-def _enrich_project_state(
+def _migrate_artifacts_table_v2(
     db: mlrun.api.db.sqldb.db.SQLDB, db_session: sqlalchemy.orm.Session
 ):
-    logger.info("Enriching projects state")
-    projects = db.list_projects(db_session)
-    for project in projects.projects:
-        changed = False
-        if not project.spec.desired_state:
-            changed = True
-            project.spec.desired_state = mlrun.common.schemas.ProjectState.online
-        if not project.status.state:
-            changed = True
-            project.status.state = project.spec.desired_state
-        if changed:
-            logger.debug(
-                "Found project without state data. Enriching",
-                name=project.metadata.name,
+    """
+    Migrate the old artifacts table to the new one, which contains the following columns:
+    uid, key, project, iteration, kind, producer_id, created, updated, and full_object
+
+    Also migrate the artifact labels and tags to the new tables
+    """
+    logger.info("Aligning artifacts")
+    new_artifacts = []
+    artifacts_tags_to_migrate = []
+    artifacts_labels_to_migrate = []
+    artifacts = db._find_artifacts(db_session, None, "*")
```

Comment:
This can have a huge memory load on the API, as regular project can contain thousands of artifacts which each can be up to 16MB, I suggest applying more defensive querying mechanism over the DB so we will go over batches rather than querying all

Reply:
Agreed - this must be batched. Can start with project-based batching, and maybe even finer-grained granularity. Of course, this means that some state will have to be kept across possible migration runs.

Reply:
> Of course, this means that some state will have to be kept across possible migration runs.

@theSaarco what do you mean by that?
What is a scenario where the migration happens in several runs?

Reply:
@TomerShor - the scenario is a migration for a setup which has many (tens of thousands, or more) of artifacts in the DB. This process can be vert lengthy and can fail at any point due to a problem, a bug, a k8s node going down or whatever else. 
In order for this to not start from scratch every time we re-migrate, it's best to keep a state of the latest point we got to. For example, if you're iterating over projects, keep the name of the last project you successfully migrated. 
The state may also be the data already in the target table - for example, if you're migrating per project, and within the project do the migration by selecting artifacts `order by id asc` then when you start you can query the target table, and see what's the last project and what's the last id in the table. Then continue from there.